include/net/net_if.h

```diff
@@ -179,6 +179,14 @@ enum net_if_flag {
 	/** Interface is in promiscuous mode */
 	NET_IF_PROMISC,
 
+	/** Do not start the interface immediately after initialization.
+	 * This requires that either the device driver or some other entity
+	 * will need to manually take the interface up when needed.
+	 * For example for Ethernet this will happen when the driver calls
+	 * the net_eth_carrier_on() function.
+	 */
+	NET_IF_NO_AUTO_START,
```

Comment:
To me this is an added feature and not the exact fix.

quoting the issue:

> boot up Zephyr without connecting network cable

It sounds like a driver/l2 issue where it enables the driver though the carrier is off which obviously should not happen. So fixing #15477 should be about this.

I am not against adding this net_if_no_auto_start though, but then its behavior could be that even carrier on would not put the if up

Reply:
> boot up Zephyr without connecting network cable

We unconditionally turn the network interface up in `net_if_post_init()`. For Ethernet this is probably not the best idea as the Ethernet driver should turn on the carrier when cable is connected which will then turn on the network interface. If the application is listening the IFACE UP event, then there is weird issues in the system if the cable is not yet connected.

Reply:
yep, that's the problem. At init phase, driver should probably check on the carrier, and set it to off/on accordingly (but not through the usual functions, since these ones will call if_up/down).

Reply:
Yes, driver needs more work, this PR is not fixing anything in that front. We just allow driver to tell that network interface is not to be started automatically.